src/providers/azure/azure.go

```diff
@@ -0,0 +1,182 @@
+		if parts[0] == "OPTION_245" && len(parts) == 2 {
+			rawEndpoint = parts[1]
+			break
+		}
+	}
+
+	if len(rawEndpoint) == 0 {
+		return nil, fmt.Errorf("DHCP option 245 not found in lease")
+	}
+
+	octets := make([]byte, 4)
+	for i := 0; i < 4; i++ {
+		octet, err := strconv.ParseUint(rawEndpoint[2*i:2*i+2], 16, 8)
+		if err != nil {
+			return nil, err
+		}
+		octets[i] = byte(octet)
+	}
+
+	return net.IPv4(octets[0], octets[1], octets[2], octets[3]), nil
+}
+
+func assertFabricCompatible(endpoint net.IP, desiredVersion string) error {
+	body, err := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+		Header: map[string][]string{
+			"x-ms-agent-name": {AgentName},
+			"x-ms-version":    {FabricProtocolVersion},
+			"Content-Type":    {"text/xml; charset=utf-8"},
+		},
+	}.Get(fmt.Sprintf("http://%s/?comp=versions", endpoint))
+	if err != nil {
+		return fmt.Errorf("failed to fetch versions: %v", err)
+	}
+
+	versions := struct {
+		Supported struct {
+			Versions []string `xml:"Version"`
+		}
+		Preferred struct {
+			Version string
+		}
+	}{}
+
+	if err := xml.Unmarshal(body, &versions); err != nil {
+		return fmt.Errorf("failed to unmarshal response: %v", err)
+	}
+
+	for _, version := range versions.Supported.Versions {
+		if version == desiredVersion {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("fabric version %s is not compatible", desiredVersion)
+}
+
+func fetchSharedConfig(endpoint net.IP) (config.Metadata, error) {
+	client := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+		Header: map[string][]string{
+			"x-ms-agent-name": {AgentName},
+			"x-ms-version":    {FabricProtocolVersion},
+			"Content-Type":    {"text/xml; charset=utf-8"},
+		},
+	}
+
+	body, err := client.Get(fmt.Sprintf("http://%s/machine/?comp=goalstate", endpoint))
+	if err != nil {
+		return config.Metadata{}, fmt.Errorf("failed to fetch goal state: %v", err)
+	}
+
+	goal := struct {
+		Container struct {
+			RoleInstanceList struct {
+				RoleInstance struct {
+					Configuration struct {
+						SharedConfig string
+					}
+				}
+			}
+		}
+	}{}
+
+	if err := xml.Unmarshal(body, &goal); err != nil {
+		return config.Metadata{}, fmt.Errorf("failed to unmarshal response: %v", err)
+	}
+
+	body, err = client.Get(goal.Container.RoleInstanceList.RoleInstance.Configuration.SharedConfig)
+	if err != nil {
+		return config.Metadata{}, fmt.Errorf("failed to fetch shared config: %v", err)
+	}
+
+	sharedConfig := struct {
+		Incarnation struct {
+			Instance string `xml:"instance,attr"`
+		}
+		Instances struct {
+			Instances []struct {
+				Id             string `xml:"id,attr"`
+				Address        string `xml:"address,attr"`
+				InputEndpoints struct {
+					Endpoints []struct {
+						LoadBalancedPublicAddress string `xml:"loadBalancedPublicAddress,attr"`
+					} `xml:"Endpoint"`
+				}
+			} `xml:"Instance"`
+		}
+	}{}
```

Comment:
It just feels icky to have these guys which amount to schema sprinkled inline with the code, but I don't know what's "idiomatic".